src/Illuminate/Database/Query/Builder.php

```diff
@@ -1995,15 +1989,50 @@ protected function enforceOrderBy()
      */
     public function pluck($column, $key = null)
     {
-        $results = $this->get(is_null($key) ? [$column] : [$column, $key]);
+        $queryResult = $this->onceWithColumns(
+            is_null($key) ? [$column] : [$column, $key],
+            function () {
+                return $this->processor->processSelect($this, $this->runSelect());
+            }
+        );
+
+        if (empty($queryResult)) {
+            return collect();
+        }
 
         // If the columns are qualified with a table or have an alias, we cannot use
         // those directly in the "pluck" operations since the results from the DB
         // are only keyed by the column itself. We'll strip the table out here.
-        return $results->pluck(
-            $this->stripTableForPluck($column),
-            $this->stripTableForPluck($key)
-        );
+        $column = $this->stripTableForPluck($column);
+        $key = $this->stripTableForPluck($key);
+
+        if (is_array($queryResult[0])) {
+            return $this->pluckFromArrayColumn($queryResult, $column, $key);
+        } else {
+            return $this->pluckFromObjectColumn($queryResult, $column, $key);
+        }
+    }
+
+    /**
+     * Execute the given callback while selecting the given columns.
+     *
+     * After running the callback, the columns are reset to the original value.
+     *
+     * @param  array     $columns
+     * @param  callable  $callback
+     * @return mixed
+     */
+    protected function onceWithColumns($columns, $callback)
+    {
+        $original = $this->columns;
+
+        if (is_null($original)) {
+            $this->columns = $columns;
+        }
+
+        return tap($callback(), function () use ($original) {
+            $this->columns = $original;
+        });
```

Comment:
I know this one is "Laravel style", but sorry it's just incredibly [painful to understand](http://www.osnews.com/story/19266/WTFs_m), especially if you are not used to this voodoo helper. Can't we just stay with the regular and simple lines:

```php
$results = $callback();

$this->columns = $original;

return $results;
```

Reply:
Yeah, you're right. (I'm used to it from Ruby land, where it's far more readable.) I've advocated against the overuse of `tap` in Laravel in the past myself, whoops. Thanks for the feedback.